### PR TITLE
`Development`: Merge exercise details calls into one DTO

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [X] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [X] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).(https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [X] I documented the Java code using JavaDoc style.

#### Client
- [X] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [X] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [X] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [X] I documented the TypeScript code using JSDoc style.

### Motivation and Context

When loading the exercise details, many different calls get executed aside of the main call. We can simplify this in one call and improve the performance.

### Description

Because converting the whole call into a full DTO is more complex and we have to make certain decisions beforehand, we decided to use a Wrapper DTO, which functions similarly to the separate calls but without the separate calls.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Course with Iris activated

1. Create a programming exercise with a submission policy
2. Create some Exercise hints by clicking on "Hints" in the exercise overview and generating some or creating them manually.
3. Participate in the exercise with all participants (one student should reach the submission policy threshold)
4. Exercise Hints should be displayed as expected.
5. Conduct plagiarism checks (make sure to choose values that create a plagiarism case
6. Once the student(s) are notified, check the student's exercise view and ensure that everything is displayed as expected: Iris, plagiarism, results.

Feel free to test more, but since I have not yet converted the exercise entity to a DTO, it contains the very same information as before (apart from an additional submission policy).

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| IrisSettingsService.java | 79.4% |                            |
| PlagiarismCaseService.java | 92.3% |               |
| ExerciseResource.java | 94.78% |               |
| PlagiarismCaseResource.java | 94.3% |               |
| exercise.service.ts | 91.71% | ✅              |
| course-exercise-details.component.ts | 98.82% | ✅              |

## Summary by CodeRabbit

- **New Features**
  - Introduced a new data transfer object (DTO) `ExerciseDetailsDTO` to encapsulate exercise details along with related settings and hints.
  - Added a method to fetch plagiarism case information for a specific exercise and user.

- **Enhancements**
  - Improved retrieval of exercise details by integrating submission policies, hints, and plagiarism case information.
  - Simplified logic for displaying minimal settings based on user roles in programming exercises.

- **Bug Fixes**
  - Corrected assignment of exercise properties in Angular components to ensure accurate data handling.

- **Refactor**
  - Updated various services and components to use `ExerciseDetailsDTO` for handling exercise details more comprehensively.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseDetailStudentParticipationDTO.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import de.tum.in.www1.artemis.domain.participation.ProgrammingExerciseStudentParticipation;
+import de.tum.in.www1.artemis.domain.participation.StudentParticipation;
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public record ExerciseDetailStudentParticipationDTO(long id, InitializationState initializationState, ZonedDateTime initializationDate, ZonedDateTime individualDueDate,
+        boolean testRun, String type, Set<ResultDTO> results, Set<SubmissionDTO> submissions, UserDTO student, TeamDTO team,
+        // Unique to programming exercises
+        String repositoryUri) {
+
+    public static ExerciseDetailStudentParticipationDTO create(StudentParticipation studentParticipation) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseDetailStudentParticipationDTO.java:28]: This method is called `of()` in other DTOs. Can we stick to that naming?

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseDetailStudentParticipationDTO.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.domain.enumeration.AssessmentType;
+import de.tum.in.www1.artemis.domain.enumeration.InitializationState;
+import de.tum.in.www1.artemis.domain.enumeration.SubmissionType;
+import de.tum.in.www1.artemis.domain.participation.ProgrammingExerciseStudentParticipation;
+import de.tum.in.www1.artemis.domain.participation.StudentParticipation;
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public record ExerciseDetailStudentParticipationDTO(long id, InitializationState initializationState, ZonedDateTime initializationDate, ZonedDateTime individualDueDate,
+        boolean testRun, String type, Set<ResultDTO> results, Set<SubmissionDTO> submissions, UserDTO student, TeamDTO team,
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseDetailStudentParticipationDTO.java:24]: Why are results and submissions included here when submissions can also contain results as well?
I know that this is the case for the entities, but what is the reason to do the same here and potentially send the same data twice?

**Review by JohannesStoehr — COMMENTED** [review]
Some intermediate feedback:

File: src/test/java/de/tum/in/www1/artemis/text/TextSubmissionIntegrationTest.java [hunk 3]
```diff
@@ -341,9 +341,9 @@
         textSubmission = ParticipationFactory.generateTextSubmission("Some text", Language.ENGLISH, true);
         textExerciseUtilService.saveTextSubmissionWithResultAndAssessor(finishedTextExercise, textSubmission, TEST_PREFIX + "student1", TEST_PREFIX + "tutor1");
 
-        Exercise returnedExercise = request.get("/api/exercises/" + finishedTextExercise.getId() + "/details", HttpStatus.OK, Exercise.class);
+        ExerciseDetailsDTO returnedExercise = request.get("/api/exercises/" + finishedTextExercise.getId() + "/details", HttpStatus.OK, ExerciseDetailsDTO.class);
 
-        assertThat(returnedExercise.getStudentParticipations().iterator().next().getResults().iterator().next().getAssessor()).as("assessor is null").isNull();
+        assertThat(returnedExercise.exercise().getStudentParticipations().iterator().next().getResults().iterator().next().getAssessor()).as("assessor is null").isNull();
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/java/de/tum/in/www1/artemis/text/TextSubmissionIntegrationTest.java:346]: It's a record

**Review by sarpsahinalp — CHANGES_REQUESTED** [review]
Client coverage seems to be failing:
Jest: "global" coverage threshold for statements (86.83%) not met: 86.82%
Jest: "global" coverage threshold for branches (73.23%) not met: 73.22%
Jest: "global" coverage threshold for lines (86.92%) not met: 86.91%
Jest: "global" coverage threshold for functions (81.24%) not met: 81.23%

**Comment by julian-christl** [comment]
> Client coverage seems to be failing: Jest: "global" coverage threshold for statements (86.83%) not met: 86.82% Jest: "global" coverage threshold for branches (73.23%) not met: 73.22% Jest: "global" coverage threshold for lines (86.92%) not met: 86.91% Jest: "global" coverage threshold for functions (81.24%) not met: 81.23%

Hey, thanks for noticing. This apparently occurred through some previous merge. I'll check it out

File: src/main/java/de/tum/in/www1/artemis/service/iris/settings/IrisSettingsService.java [hunk 4]
```diff
@@ -401,6 +406,17 @@ public IrisCombinedSettingsDTO getCombinedIrisSettingsFor(Exercise exercise, boo
...
     }
 
+    /**
+     * Check if we have to show minimal settings for an exercise. Editors can see the full settings, students only the reduced settings.
+     *
+     * @param exercise The exercise to check
+     * @param user     The user to check
+     * @return Whether we have to show the user the minimal settings
+     */
+    public boolean showMinimalSettings(Exercise exercise, User user) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/iris/settings/IrisSettingsService.java:416]:
```suggestion
    public boolean shouldShowMinimalSettings(Exercise exercise, User user) {
```

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/PlagiarismCaseService.java [hunk 5]
```diff
@@ -115,6 +117,26 @@ public void createOrAddToPlagiarismCasesForComparison(long plagiarismComparisonI
...
+     * @return the plagiarism case for the student and exercise if it exists
+     */
+    public Optional<PlagiarismCaseInfoDTO> getPlagiarismCaseInfoForExerciseAndUser(long exerciseId, long userId) {
+        return plagiarismCaseRepository.findByStudentIdAndExerciseIdWithPost(userId, exerciseId)
+                // the student was notified if the plagiarism case is available (due to the nature of the query above)
+                // the following line is already checked in the SQL statement, but we want to ensure it 100%
+                .filter((plagiarismCase) -> plagiarismCase.getPost() != null).map((plagiarismCase) -> {
+                    // Note: we only return the ID and verdict to tell the client there is a confirmed plagiarism case with student notification (post) and to support navigating to
+                    // the
+                    // detail page
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/PlagiarismCaseService.java:134]:
```suggestion
                    // the detail page
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java [hunk 6]
```diff
@@ -321,19 +338,24 @@ public ResponseEntity<Exercise> getExerciseDetails(@PathVariable Long exerciseId
...
         }
 
-        return ResponseEntity.ok(exercise);
+        IrisCombinedSettingsDTO irisSettings = irisSettingsService.map(service -> service.getCombinedIrisSettingsFor(exercise, service.showMinimalSettings(exercise, user)))
+                .orElse(null);
+        PlagiarismCaseInfoDTO plagiarismCaseInfo = plagiarismCaseService.getPlagiarismCaseInfoForExerciseAndUser(exercise.getId(), user.getId()).orElse(null);
+
+        if (exercise instanceof ProgrammingExercise programmingExercise) {
+            Set<ExerciseHint> activatedExerciseHints = exerciseHintService.getActivatedExerciseHints(programmingExercise, user);
+            Set<ExerciseHint> availableExerciseHints = exerciseHintService.getAvailableExerciseHints(programmingExercise, user);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:354]: This method uses a lot of database calls and as far as I can tell, we don't check first if there are any hints available. So we should probably wrap this into an if-statement that first checks if we actually need these database calls.
One option would be to load the exercise hints in the method first and check if there are any

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseDetailsDTO.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+import de.tum.in.www1.artemis.domain.Exercise;
+import de.tum.in.www1.artemis.domain.hestia.ExerciseHint;
+import de.tum.in.www1.artemis.service.iris.dto.IrisCombinedSettingsDTO;
+import de.tum.in.www1.artemis.web.rest.dto.plagiarism.PlagiarismCaseInfoDTO;
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public record ExerciseDetailsDTO(Exercise exercise, IrisCombinedSettingsDTO irisSettings, PlagiarismCaseInfoDTO plagiarismCaseInfo, Set<ExerciseHint> availableExerciseHints,
+        Set<ExerciseHint> activatedExerciseHints) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseDetailsDTO.java:14]: Why do we not convert the ExerciseHints into DTOs as well?

**Review by JohannesStoehr — DISMISSED** [review]

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseDetailsDTO.java (hunk 7 above)
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseDetailsDTO.java:14]: This PR doesn't contain any conversion. We only use a wrapper DTO. The conversion of the Exercise model to a DTO got scrapped for now.

**Comment by julian-christl** [comment]
@coderabbitai pause

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/PlagiarismCaseService.java [hunk 8]
```diff
@@ -115,6 +117,25 @@ public void createOrAddToPlagiarismCasesForComparison(long plagiarismComparisonI
...
+     * @param userId     the ID of the student
+     * @return the plagiarism case for the student and exercise if it exists
+     */
+    public Optional<PlagiarismCaseInfoDTO> getPlagiarismCaseInfoForExerciseAndUser(long exerciseId, long userId) {
+        return plagiarismCaseRepository.findByStudentIdAndExerciseIdWithPost(userId, exerciseId)
+                // the student was notified if the plagiarism case is available (due to the nature of the query above)
+                // the following line is already checked in the SQL statement, but we want to ensure it 100%
+                .filter((plagiarismCase) -> plagiarismCase.getPost() != null).map((plagiarismCase) -> {
+                    // Note: we only return the ID and verdict to tell the client there is a confirmed plagiarism case with student notification (post) and to support navigating to
+                    // the detail page
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/PlagiarismCaseService.java:133]:
```suggestion
                    // Note: we only return the ID and verdict to tell the client there is a confirmed plagiarism case with student notification (post)
                    // and to support navigating to the detail page
```

File: src/main/webapp/app/exercises/shared/exercise/exercise.service.ts [hunk 9]
```diff
@@ -146,15 +158,18 @@ export class ExerciseService {
...
+            map((res: EntityDetailsResponseType) => {
                 if (res.body) {
+                    res.body.exercise = ExerciseService.convertExerciseDatesFromServer(res.body.exercise)!;
+                    ExerciseService.parseExerciseCategories(res.body.exercise);
                     // insert an empty list to avoid additional calls in case the list is empty on the server (because then it would be undefined in the client)
-                    if (res.body.posts === undefined) {
-                        res.body.posts = [];
+                    if (res.body.exercise.posts === undefined) {
+                        res.body.exercise.posts = [];
+                    }
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/exercise/exercise.service.ts:170]:
```suggestion
                    res.body.exercise.posts ??= [];
```

**Review by Strohgelaender — COMMENTED** [review]
code lgtm, just two small nitpicks.

File: src/main/webapp/app/exercises/shared/exercise/exercise.service.ts (hunk 9 above)
Comment (julian-christl) [reply src/main/webapp/app/exercises/shared/exercise/exercise.service.ts:170]: Honestly, I had no idea that this was possible before and I also don't really think it's too intuitive. An optional assign...

I'd prefer to keep it this way if it's okay for you. If we use it throughout the codebase and I just didn't realise it until now, let me know and we can change this instance too.

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by az108 — COMMENTED** [review]
I just noticed this but why doesnt the hint dissapear after it was activated. And pressing it again only leads to errors.
image

**Review by az108 — COMMENTED** [review]
Works as expected. Tested it on TS3
image

**Review by az108 — APPROVED** [review]
Code LGTM aswell

**Review by BBesrour — APPROVED** [review]
Tested on TS1, works as expected

**Review by laadvo — APPROVED** [review]
Code LGTM

**Pull request merged**